### ISSUE TITLE
JSR250 support for web-flux

### ISSUE THREAD

**Issue by bdeneuter** [spring-projects/spring-security#5103]
### Summary

I'm migrating an MVC application to web-flux. Spring security supports the annotations of JSR250 (RolesAllowed, ...) for Spring MVC applications but not for web-flux applications.

[https://docs.spring.io/spring-security/site/docs/5.0.3.RELEASE/reference/htmlsingle/#jc-method](url)

**Comment by rwinch** [comment]
You must use `@EnableReactiveMethodSecurity` for reactive applications. You also need to ensure that your methods all return either `Mono` or `Flux` (this is the way Reactor Context propagates the `SecurityContext`. Currently `@EnableReactiveMethodSecurity` will work for `@PreAuthorize` and `@PostAuthorize` annotations which support a super set of the `JSR` annotations. You can find a [sample](https://github.com/spring-projects/spring-security/tree/5.0.3.RELEASE/samples/javaconfig/hellowebflux-method) in Spring Security's samples.

However, JSR annotations are not yet supported. I have scheduled adding JSR based annotations for Spring Security 5.1.

**Comment by Ernir** [comment]
I am currently running Spring Security version 5.1.4, and this does not seem to be fully resolved. I see a `Jsr250MethodSecurityMetadataSource` bean, but the issue is effectively unchanged. Adding `@EnableGlobalMethodSecurity(jsr250Enabled = true)` to a configuration class that already has `@EnableReactiveMethodSecurity` now results in _three_ beans being found:

```
     | Parameter 0 of method securityMethodInterceptor in org.springframework.security.config.annotation.method.configuration.ReactiveMethodSecurityConfiguration required a single bean, but 3 were found:
     |        - methodMetadataSource: defined by method 'methodMetadataSource' in class path resource [org/springframework/security/config/annotation/method/configuration/ReactiveMethodSecurityConfiguration.class]
     |        - methodSecurityMetadataSource: defined by method 'methodSecurityMetadataSource' in class path resource [org/springframework/security/config/annotation/method/configuration/GlobalMethodSecurityConfiguration.class]
     |        - jsr250MethodSecurityMetadataSource: defined by method 'jsr250MethodSecurityMetadataSource' in class path resource [org/springframework/security/config/annotation/method/configuration/Jsr250MetadataSourceConfiguration.class]
```

Was this fixed in 5.1?

**Comment by jzheaux** [comment]
@Ernir no, there isn't Reactive support for JSR annotations yet - this ticket will remain open until that support is added.

**Comment by Ernir** [comment]
Thank you for your reply, @jzheaux. Can you tell me if reactive support for the annotations is scheduled for a particular version?

I am also interested in the original `@Secured` annotation.

**Comment by jzheaux** [comment]
I believe this ticket is tracking the Reactive JSR250 (@Secured) support, so once it's scheduled, you'll see a milestone selected in the right-hand column of the ticket.

If you would like, I'd be happy to help you get a PR going to add the support. Would you be interested in doing that?

**Comment by MrJovanovic13** [comment]
Hi. May I work on this issue? @jzheaux 

I would appreciate any pointers I can get. I took a look at the implementation of `Jsr250AuthorizationManager`. 
I assume I would need to implement a reactive variant? Thanks for any pointers!

**Comment by celikfatih** [comment]
If no one is currently working on this issue, I’d be happy to take a look.

From what I understand, **JSR-250** annotations like `@RolesAllowed`, `@PermitAll`, and `@DenyAll` are supported in the servlet stack but not currently integrated with reactive method security.

My plan would be to explore how`Jsr250MethodSecurityMetadataSource` is wired in the servlet method security configuration and introduce a similar integration for reactive method security so that these annotations can work with `@EnableReactiveMethodSecurity`. I would also add tests demonstrating their behavior in the reactive stack.

Please let me know if this approach aligns with what you had in mind. If so, I can start working on a PR.

Thanks
